tbdex/rfq/create.go

```diff
@@ -0,0 +1,220 @@
+package rfq
+
+import (
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	"github.com/TBD54566975/tbdex-go/tbdex"
+	"github.com/tbd54566975/web5-go/crypto"
+	"go.jetpack.io/typeid"
+)
+
+// Create creates an [RFQ]
+//
+// # An RFQ is a resource created by a customer of the PFI to request a quote
+//
+// [RFQ]: https://github.com/TBD54566975/tbdex/tree/main/specs/protocol#rfq-request-for-quote
+func Create(from, to, offeringID string, payin PayinMethodWithDetails, payout PayoutMethodWithDetails, opts ...CreateOption) (RFQ, error) {
+	r := createOptions{
+		id:         typeid.Must(typeid.WithPrefix(RFQKind)).String(),
+		createdAt:  time.Now(),
+		protocol:   "1.0",
+		externalID: "",
+	}
+
+	for _, opt := range opts {
+		opt(&r)
+	}
+
+	hashedData, privateData, err := hashData(&payin, &payout, r.claims)
+	if err != nil {
+		return RFQ{}, fmt.Errorf("failed to hash private data: %w", err)
+	}
+
+	return RFQ{
+		MessageMetadata: tbdex.MessageMetadata{
+			From:       from,
+			To:         to,
+			Kind:       RFQKind,
+			ID:         r.id,
+			ExchangeID: r.id,
+			CreatedAt:  r.createdAt.UTC().Format(time.RFC3339),
+			ExternalID: r.externalID,
+			Protocol:   r.protocol,
+		},
+		Data: RFQData{
+			OfferingID: offeringID,
+			Payin: SelectedPayinMethod{
+				Amount:             payin.Amount,
+				Kind:               payin.Kind,
+				PaymentDetailsHash: hashedData.PayinHash,
+			},
+			Payout: SelectedPayoutMethod{
+				Kind:               payin.Kind,
+				PaymentDetailsHash: hashedData.PayoutHash,
+			},
+			ClaimsHash: hashedData.ClaimsHash,
+		},
+		PrivateData: &privateData,
+	}, nil
+}
+
+func hashData(payin *PayinMethodWithDetails, payout *PayoutMethodWithDetails, claims []string) (rfqHashes, RFQPrivateData, error) {
+	randomBytes, err := crypto.GenerateEntropy(crypto.Entropy128)
+	if err != nil {
+		return rfqHashes{}, RFQPrivateData{}, err
+	}
+
+	salt := base64.URLEncoding.EncodeToString(randomBytes)
+
+	payinHash, err := digestData(salt, payin)
```

Comment:
If no payment details are passed, i.e. paymentDetails is null, then we should also set payinHash to null. This also goes for payout. I think existing implementations handle claims the same way, but looks like I left a discrepancy in the spec. Gimme a sec to double check the existing implementations and decide whether we need to update the spec for null/empty claims.

Reply:
spec change PR to clarify this in the spec: https://github.com/TBD54566975/tbdex/pull/313